### DISCUSSION TITLE
Accessing a cell programatically

### DISCUSSION THREAD

**Discussion by juancrl** [dotnet/interactive#2075] [Q&A]
Which is the .NET object(s) to use to access the Cell object ? Preferably from PowerShell.

   ( I mean, in run time, not reading the static *.ipynb file and converting to Json, as @dfinke is doing in his fantastic module.)

Specifically, I need to get/set the cell WIDTH, clearing all output, changing colors (everything we had in $HOST.ui...), 
and also minimizing (collapsing) a cell, enumerating them... 

Any link to the documentation will be enough; a short one-liner example also welcome.

TIA,
jc

**Comment by jonsequitur** [comment]
The kernel itself doesn't know anything about the file or the UI, so the answer to your question would be in the VS Code APIs (if you're using .NET Interactive via the notebooks extension) or in the various Jupyter frontend APIs (e.g. [Jupyter Lab extensions](https://jupyterlab.readthedocs.io/en/stable/user/extensions.html)). 

For this  reason, accessing these from code within a notebook cell would be highly specific to the specific frontend.

The contents of cell abouts are a little more accessible via HTML, CSS, and JavaScript, though.

<img width="622" alt="image" src="https://user-images.githubusercontent.com/547415/173875505-3a067460-ed9f-48ba-9aca-f397251e23a8.png">

**Reply by jonsequitur** [reply]
You can also do this programmatically:

<img width="544" alt="image" src="https://user-images.githubusercontent.com/547415/173876601-93c0a27e-32cf-455d-aa64-9975ebf0fcc6.png">

**Reply by juancrl** [reply]
Thanks Jon for your help, effort and patience. 

I don't know if this is the best channel for questions (yes, it is, I'm getting some answers at last)...

I'm a hardcore PowerShell fan... I was working in Microsoft when Snover/Payette published the first internal beta of monad in \\AXP... and have built some code since then... almost 20 years of PowerShell.

For me this environment is a bit strange. There's a PowerShell host running by someone (VSCode ? Jupyter or .NET Interactive addons ? Python executable from it ?), with a call and setup chain I don't understand (Anaconda ? Python ? )... at least my environment reverts to that sometimes. 

And the worst part is the $Host.UI.RawUi methods are not callable from the code. Clear-Host does not work. (Even VsCode has an option to clear all cells),,, and being all .NET, I assume there are some objects handling Cells and Outputs and Hosts, maybe non-visible and needing reflection, but I cannot find them.


Calling Javascript will work, and thanks for the hint... I'll look for [Kernel.root] to call from PowerShell and see other options, as how to collapse the cell, or set a bigger output width...

**Reply by jonsequitur** [reply]
The .NET portion of the code (running in the `dotnet-interactive` executable, i.e. the kernel) has no concept of cells whatsoever. The UI specifics are all implemented in TypeScript and are running in the VS Code extension host. You can think of the notebook as a bit like a single-page web app, where the kernel is the server and the UI is the rendered web page. Almost all UI logic is in the frontend, not the server. The programmatic model for that frontend varies. VS Code and Jupyter Lab have entirely different models which the kernel is intentionally decoupled from.

**Reply by juancrl** [reply]
Thanks again. I've found a way to make the usable host (console) width broader, I'll investigate the other options. 
But it seems evident the PowerShell $host is incomplete in this environment, it does not offer all the functionality. 
Just the capacity to clear its own cell would be basic (or maybe it exists and just it's not evident).
